llvm/docs/Telemetry.rst

```diff
@@ -70,120 +69,168 @@ Key components
+    Entry->SessionId = SessionId;
+    emitToAllDestinations(Entry);
+  }
       
-      json::Object serializeToJson() const {
-        json::Object Ret = TelemeteryInfo::serializeToJson();
-        Ret.emplace_back("MyMsg", Msg);
-        return std::move(Ret);
+  void addDestination(std::unique_ptr<Destination> dest) override {
+    destinations.push_back(std::move(dest));
+  }
+  
+  // You can also define additional instrumentation points.
+  void logStartup(TelemetryInfo* Entry) {
+    // Add some additional data to entry.
+    Entry->Msg = "Some message";
+    dispatch(Entry);
+  }
+  
+  void logAdditionalPoint(TelemetryInfo* Entry) {
+    // .... code here
+  }
+  
+  private:
+    void emitToAllDestinations(const TelemetryInfo* Entry) {
+      for (Destination* Dest : Destinations)
+        Dest->receiveEntry(Entry);
       }
-      
-      // TODO: implement getKind() and classof() to support dyn_cast operations.
-    };
+    }
     
-    class MyTelemeter : public llvm::telemery::Telemeter {
-    public:
-      static std::unique_ptr<MyTelemeter> createInstatnce(llvm::telemetry::Config* config) {
-        // If Telemetry is not enabled, then just return null;
-        if (!config->EnableTelemetry) return nullptr;
-        
-        std::make_unique<MyTelemeter>();
-      }
-      MyTelemeter() = default;
-      
-      void logStartup(llvm::StringRef ToolName, TelemetryInfo* Entry) override {
-        if (MyTelemetryInfo* M = dyn_cast<MyTelemetryInfo>(Entry)) {
-          M->Msg = "Starting up tool with name: " + ToolName;
-          emitToAllDestinations(M);
-        } else {
-          emitToAllDestinations(Entry);
-        }
+    std::vector<Destination> Destinations;
+    const std::string SessionId;
+  };
+
+  class MyDestination : public telemetry::Destination {
+  public:
+    Error receiveEntry(const TelemetryInfo* Entry) override {
+      if (Error err = serializer.start()) {
+        return err;
       }
-      
-      void logExit(llvm::StringRef ToolName, TelemetryInfo* Entry) override {
-        if (MyTelemetryInfo* M = dyn_cast<MyTelemetryInfo>(Entry)) {
-          M->Msg = "Exitting tool with name: " + ToolName;
-          emitToAllDestinations(M);
-        } else {
-          emitToAllDestinations(Entry);
-        }
-      }
-      
-      void addDestination(Destination* dest) override {
-        destinations.push_back(dest);
+      Entry->serialize(serializer);
+      if (Error err = serializer.finish()) {
+        return err;
       }
+
+      json::Object copied = *serializer.getOutputObject();
+      // Send the `copied` object to wherever.
+      return Error::success();
+    }
+
+  private:
+    JsonSerializer serializer;
+  };
+
+  // This defines a custom TelemetryInfo that has an addition Msg field.
+  struct MyTelemetryInfo : public telemetry::TelemetryInfo {
+    std::chrono::time_point<std::chrono::steady_clock> Start;
+    std::chrono::time_point<std::chrono::steady_clock> End;
```

Comment:
As these aren't used in the serialize method, perhaps they should be deleted/moved/...?

Reply:
done